staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/objectmeta/validation.go

```diff
@@ -0,0 +1,123 @@
+func Validate(pth *field.Path, obj interface{}, s *structuralschema.Structural, includeRoot bool) field.ErrorList {
+	if includeRoot {
+		if s == nil {
+			s = &structuralschema.Structural{}
+		}
+		clone := *s
+		clone.XEmbeddedResource = true
+		s = &clone
+	}
+	return validate(pth, obj, s)
+}
+
+func validate(pth *field.Path, x interface{}, s *structuralschema.Structural) field.ErrorList {
+	if s == nil {
+		return nil
+	}
+
+	var allErrs field.ErrorList
+
+	switch x := x.(type) {
+	case map[string]interface{}:
+		if s.XEmbeddedResource {
+			allErrs = append(allErrs, validateEmbeddedResource(pth, x, s)...)
+		}
+
+		for k, v := range x {
+			prop, ok := s.Properties[k]
+			if ok {
+				allErrs = append(allErrs, validate(pth.Child(k), v, &prop)...)
+			} else if s.AdditionalProperties != nil {
+				allErrs = append(allErrs, validate(pth.Key(k), v, s.AdditionalProperties.Structural)...)
+			}
+		}
+	case []interface{}:
+		for i, v := range x {
+			allErrs = append(allErrs, validate(pth.Index(i), v, s.Items)...)
+		}
+	default:
+		// scalars, do nothing
+	}
+
+	return allErrs
+}
+
+func validateEmbeddedResource(pth *field.Path, x map[string]interface{}, s *structuralschema.Structural) field.ErrorList {
+	var allErrs field.ErrorList
+
+	// require apiVersion and kind, but not metadata
+	if s.XEmbeddedResource {
+		if _, found := x["apiVersion"]; !found {
+			allErrs = append(allErrs, field.Required(pth.Child("apiVersion"), "must not be empty"))
+		}
+		if _, found := x["kind"]; !found {
+			allErrs = append(allErrs, field.Required(pth.Child("kind"), "must not be empty"))
+		}
+	}
+
+	for k, v := range x {
+		if s.XEmbeddedResource {
+			switch k {
+			case "apiVersion":
+				if apiVersion, ok := v.(string); !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, "must be a string"))
+				} else if len(apiVersion) == 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, "must not be empty"))
+				} else if _, err := schema.ParseGroupVersion(apiVersion); err != nil {
+					allErrs = append(allErrs, field.Invalid(pth.Child("apiVersion"), apiVersion, err.Error()))
+				}
+			case "kind":
+				if kind, ok := v.(string); !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), v, "must be a string"))
+				} else if len(kind) == 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), kind, "must not be empty"))
+				} else if errs := utilvalidation.IsDNS1035Label(strings.ToLower(kind)); len(errs) > 0 {
+					allErrs = append(allErrs, field.Invalid(pth.Child("kind"), kind, "may have mixed case, but should otherwise match: "+strings.Join(errs, ",")))
+				}
+			case "metadata":
+				meta, _, err := GetObjectMeta(x, true)
+				if err != nil {
+					allErrs = append(allErrs, field.Invalid(pth.Child("metadata"), v, err.Error()))
+				} else {
+					if len(meta.Name) == 0 {
+						meta.Name = "fakename" // we have to set something to avoid an error
```

Comment:
I had some trouble following this part. If ValidateObjectMeta expects name generation to have been performed, by setting a fake name here do we end up not ever doing the name validation as intended once name generation is performed? Or is that handled elsewhere?

Reply:
name generation is generally done before validation. Not sure if that holds true for custom resources as well

Reply:
https://github.com/kubernetes/kubernetes/blob/c7972d9a5e7f555371fec7a3237f08ebc1ae9d7c/staging/src/k8s.io/apiserver/pkg/registry/rest/create.go#L91-L110

Reply:
The main point here is that it is totally fine to have an object without name in a CustomResource.

Reply:
added test for "no name"